File: scripts/export-training-pack.mjs
import fs from "fs";
import path from "path";
import AdmZip from "adm-zip";

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) continue;
    const key = item.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      out[key] = true;
      continue;
    }
    out[key] = next;
    i += 1;
  }
  return out;
}

function timestamp() {
  const d = new Date();
  const p = (n) => String(n).padStart(2, "0");
  return `${d.getFullYear()}${p(d.getMonth() + 1)}${p(d.getDate())}-${p(d.getHours())}${p(d.getMinutes())}${p(d.getSeconds())}`;
}

function addPathIfExists(zip, rootDir, relPath, fileList) {
  const full = path.resolve(rootDir, relPath);
  if (!fs.existsSync(full)) return;
  const stat = fs.statSync(full);
  if (stat.isFile()) {
    zip.addLocalFile(full, path.dirname(relPath).replace(/\\/g, "/"));
    fileList.push(relPath.replace(/\\/g, "/"));
    return;
  }
  if (stat.isDirectory()) {
    zip.addLocalFolder(full, relPath.replace(/\\/g, "/"));
    fileList.push(`${relPath.replace(/\\/g, "/")}/`);
  }
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const includeEnv = String(args["include-env"] || "false").toLowerCase() === "true";
  const outDir = path.resolve(process.cwd(), "backups");
  if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });

  const outFile =
    args.output
      ? path.resolve(String(args.output))
      : path.join(outDir, `training-pack-${timestamp()}.zip`);

  const zip = new AdmZip();
  const root = process.cwd();
  const fileList = [];
  const targets = [
    "config/profiles",
    "data/training",
    "models"
  ];
  if (includeEnv) targets.push(".env");

  for (const t of targets) {
    addPathIfExists(zip, root, t, fileList);
  }

  const manifest = {
    createdAt: new Date().toISOString(),
    sourceHost: process.env.COMPUTERNAME || process.env.HOSTNAME || "unknown",
    includeEnv,
    files: fileList
  };
  zip.addFile("training-pack.manifest.json", Buffer.from(JSON.stringify(manifest, null, 2), "utf8"));
  zip.writeZip(outFile);

  console.log(`Exported: ${outFile}`);
  console.log(`Entries: ${fileList.length}`);
}

main();

File: scripts_py/export_training_pack.py
import os
import json
import zipfile
from datetime import datetime, timezone
from pathlib import Path

from common_pipeline import parse_args, to_bool


def _ts():
    return datetime.now().strftime("%Y%m%d-%H%M%S")


def _add_path(zf, root, rel, files):
    p = (root / rel).resolve()
    if not p.exists():
        return
    if p.is_file():
        zf.write(p, arcname=str(Path(rel).as_posix()))
        files.append(str(Path(rel).as_posix()))
        return
    for fp in p.rglob("*"):
        if not fp.is_file():
            continue
        arc = str(fp.resolve().relative_to(root).as_posix())
        zf.write(fp, arcname=arc)
    files.append(f"{Path(rel).as_posix()}/")


if __name__ == "__main__":
    args = parse_args(os.sys.argv[1:])
    root = Path(os.getcwd()).resolve()
    out_dir = root / "backups"
    out_dir.mkdir(parents=True, exist_ok=True)
    out_file = Path(str(args.get("output") or (out_dir / f"training-pack-{_ts()}.zip"))).resolve()
    include_env = to_bool(args.get("include-env"), False)
    sqlite_only = to_bool(args.get("sqlite-only"), False)
    targets = ["config/profiles", "data/training/training.db", "models"] if sqlite_only else ["config/profiles", "data/training", "models"]
    if include_env:
        targets.append(".env")
    files = []
    with zipfile.ZipFile(out_file, "w", compression=zipfile.ZIP_DEFLATED) as zf:
        for rel in targets:
            _add_path(zf, root, rel, files)
        manifest = {
            "createdAt": datetime.now(timezone.utc).isoformat(),
            "sourceHost": os.environ.get("COMPUTERNAME") or os.environ.get("HOSTNAME") or "unknown",
            "includeEnv": include_env,
            "sqliteOnly": sqlite_only,
            "files": files,
        }
        zf.writestr("training-pack.manifest.json", json.dumps(manifest, indent=2, ensure_ascii=False))
    print(f"Exported: {out_file}")
    print(f"Entries: {len(files)}")
    print(f"SQLite only: {str(sqlite_only).lower()}")
